ValveResourceFormat/IO/HammerMeshBuilder.cs

```diff
@@ -0,0 +1,1021 @@
+                HalfEdge boundaryHe = null;
+
+                if (v1.outGoingHalfEdge != -1)
+                {
+                    var hasBoundary = false;
+                    foreach (var he in v1.relatedEdges)
+                    {
+                        if (HalfEdges[he].face == -1)
+                        {
+                            hasBoundary = true;
+                        }
+                    }
+                    if (!hasBoundary)
+                    {
+                        halfEdgeModifier.RollBack($"Removed face `{OriginalFaceCount - FacesRemoved}`, Face specified a vertex which had edges attached, but none that were open.");
+                        return;
+                    }
+                }
+
+                if (v2.outGoingHalfEdge != -1)
+                {
+                    var hasBoundary = false;
+                    foreach (var he in v2.relatedEdges)
+                    {
+                        if (HalfEdges[he].face == -1)
+                        {
+                            hasBoundary = true;
+                        }
+                    }
+                    if (!hasBoundary)
+                    {
+                        halfEdgeModifier.RollBack($"Removed face `{OriginalFaceCount - FacesRemoved}`, Face specified a vertex which had edges attached, but none that were open.");
+                        return;
+                    }
+                }
+
+                // check if the inner already exists
+                if (!halfEdgeModifier.TryAddVertsToEdgeDict(innerHeKey, innerHeIndex))
+                {
+                    // failed to add the key to the dict, this either means we hit a set of inner twins
+                    // or that the face being added is wrong
+
+                    VertsToEdgeDict.TryGetValue(innerHeKey, out innerHeIndex); // get the half edge that already exists and assign to innerHeIndex
+
+                    innerHe = HalfEdges[innerHeIndex];
+
+                    // already existing half edge doesn't have a face, which means this is a boundary, don't add a new half edge
+                    // but instead just set its face to be the current face (turning it into an inner)
+                    if (innerHe.face == -1)
+                    {
+                        if (lastinnerFaceHeidx != -1)
+                        {
+                            if (HalfEdges[lastinnerFaceHeidx].overridOuter)
+                            {
+                                if (innerHe.prev != lastinnerFaceHeidx && HalfEdges[lastinnerFaceHeidx].next != innerHeIndex)
+                                {
+                                    halfEdgeModifier.RollBack($"Removed face `{OriginalFaceCount - FacesRemoved}`, Face specified two edges that are connected by a vertex but have or more existing edges separating them.");
+                                    return;
+                                }
+                            }
```

Comment:
The nesting is excessive. Extract boolean checks to variables with descriptive names and group them.